Implement struct serializer

Baryonics opened the pull request.

YanzhaoW commented:
Ok, #31 is merged. You could start to implement the class (make sure you have updated your local master branch).

First change `EXPECT_FALSE` to `EXPECT_TRUE` here:

https://github.com/YanzhaoW/srs-control/blob/8de00f58c23519e8a53413df26309c3d7e3a1dfd/test/backend/srs/UnitTestStruct.cpp#L83-L87

Then, implement the convert function here to make sure the test succeeds.

https://github.com/YanzhaoW/srs-control/blob/8de00f58c23519e8a53413df26309c3d7e3a1dfd/backend/srs/converters/StructSerializer.cpp#L17-L23

To do the testing, go to the `build/bin` folder and run

```bash
./unit_test_srs_backend
```

Baryonics committed:
Move *CompactData structs to SRSDataCompact.hpp
Serializer/Deserializer now share MarkerDataCompact and HitDataCompact.
Add converters between *DataStruct and *CompactDataStruct and add
bin_to_gray helper.

YanzhaoW commented on backend/srs/data/SRSDataCompact.hpp:
I believe this file has not been not formatted correctly.

YanzhaoW commented on backend/srs/data/CMakeLists.txt:
```diff
@@ -15,6 +15,7 @@ target_sources(
            FILES
            SRSDataStructs.hpp
            DataStructsFormat.hpp
+           SRSDataCompact.hpp  # is this it? 
```
We could add it here.

YanzhaoW commented on backend/srs/utils/CommonFunctions.hpp:
```diff
@@ -50,6 +51,21 @@ namespace srs::common
+        for (auto j{low_size}; j < high_size; j++)
+        {
+            high_bits[j] = bits[low_size + j];
+        }
```
Maybe there is a better alternative to this raw looping.

YanzhaoW reviewed (commented):
Thanks for the PR. Here are some suggestions/comments for now:

YanzhaoW commented on backend/srs/utils/CommonFunctions.hpp:
```diff
@@ -71,6 +87,12 @@ namespace srs::common
     }
 
+    template<typename T>
+    constexpr auto bin_to_gray(T bin_val)
```
Maybe use `binary_to_gray` to be more consistent with `gray_to_binary`.

Baryonics committed:
Made changes according to suggestions.

bin_to_gray -> binary_to_gray, avoided raw for loops in split_bits,
formatting with clangd.format

Implemented a covert function

Changed order of includes to satisfy clang-format

Get rid of magic numbers

fixed wrong required size

Implemented the convert function
Bug fixes on binary_to_gray function

Baryonics committed:
Added monoadic conditions for safety.
Readded test case that should have prevented a wrong sized bitset

YanzhaoW commented on backend/srs/converters/StructSerializer.cpp:
```diff
@@ -1,13 +1,108 @@
+            auto compact_bitset =
+                std::bitset<sizeof(std::uint64_t) * common::BYTE_BIT_LENGTH>{ std::bit_cast<uint64_t>(compact_data) };
+            auto output = std::vector<char>{};
+            output.resize(sizeof(std::uint64_t) / common::BYTE_BIT_LENGTH);
```
Well, this is not a good decision to create a new vector in a loop. When a new vector is created, you have the memory allocation (in heap), which is very slow and should be avoided.

What you could do is to create some buffer vector and initialize it only once in the constructor. In the loop, before you write something into it, reset all its members to be zero. Your `compact_to_vector` should take an non-const ref as its output:

```cpp
void compact (const T& compact_data, std::vector<char>& output);
```

YanzhaoW replied:
You sure `output.resize(sizeof(std::uint64_t) / common::BYTE_BIT_LENGTH);` is correct? `sizeof(std::uint64_t) / common::BYTE_BIT_LENGTH` always results in 1.

YanzhaoW reviewed (commented):
Looks great! Here are some further suggestions:

Baryonics committed:
Fixed wrong size

Fixed a few bugs and tests pass now

Avoid vector creation in a loop

Fixed UB that causes clang to fail ctest.

Fixe a few bugs. Removed vector creation from loop.

YanzhaoW commented on backend/srs/converters/StructSerializer.cpp:
```diff
@@ -18,9 +114,32 @@ namespace srs::process
-        return std::unexpected{ "not implemented" };
+        auto serialize_to = zpp::bits::out{ output, zpp::bits::endian::network{}, zpp::bits::no_size{} };
+        serialize_to(input->header).or_throw();
+        auto output_buffer = std::vector<char>{};
```
This is still a problem. Since the convert function is called for each event, you have memory allocations multiple times.

The call site of this convert function is here:

https://github.com/YanzhaoW/srs-control/blob/1e2726a3b9476b36f4c127bbaabccf0b4b0b7405/backend/srs/converters/StructSerializer.hpp#L22-L26 

A better solution is to make this output_buffer as a member variable (for each pipeline). In the beginning of the convert function, clear the vector.

Baryonics committed:
splittet split_bit function into two function instead of returning a pair

Baryonics committed:
Implemented a buffer member variable

YanzhaoW commented on backend/srs/converters/StructSerializer.hpp:
```diff
@@ -32,5 +32,6 @@ namespace srs::process
 
       private:
         std::vector<std::vector<char>> output_data_;
+        std::vector<char> compact_data_buffer_;
```
There will be data race problem in the multithreading mode. Instead of just one buffer for all lines, create a buffer for each line (just like `output_data_`):
```cpp
std::vector<std::vector<char>> compact_data_buffers_;
```
Then in the constructor, initialize them with the number of lines. Please check how the `output_data_` is dealt with.

Baryonics committed:
Prevent data race
